syntax = "proto3";
package phoenix.proto.trace.v1;

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";
import "google/protobuf/wrappers.proto";

message Span {
  google.protobuf.Struct attributes = 1;
  message Context {
    string trace_id = 1;
    string span_id = 2;
  }
  Context context = 2;
  google.protobuf.StringValue parent_span_id = 3;
  string name = 4;
  google.protobuf.Timestamp start_time = 5;
  google.protobuf.Timestamp end_time = 6;
  message Status {
    enum Code {
      UNSET = 0;
      OK = 1;
      ERROR = 2;
    };
    Code code = 1;
    google.protobuf.StringValue message = 2;
  }
  Status status = 7;
  message Event {
    google.protobuf.Struct attributes = 1;
    string name = 2;
    google.protobuf.Timestamp timestamp = 3;
  }
  repeated Event events = 8;
  message Exception {
    google.protobuf.Struct attributes = 1;
    google.protobuf.StringValue message = 2;
    google.protobuf.StringValue type = 3;
    google.protobuf.BoolValue escaped = 4;
    google.protobuf.StringValue stacktrace = 5;
    google.protobuf.Timestamp timestamp = 6;
  }
  repeated Exception exceptions = 9;
  message IOValue {
    oneof kind {
      string string_value = 1;
      google.protobuf.Struct json_value = 2;
    }
  }
  IOValue input = 10;
  IOValue output = 11;
  string kind = 12;
  Retrieval retrieval = 13;
  Embedding embedding = 14;
  LLM llm = 15;
  Tool tool = 16;
}
message Retrieval {
  message Document {
    google.protobuf.Struct attributes = 1;
    google.protobuf.StringValue id = 2;
    google.protobuf.FloatValue score = 3;
    google.protobuf.StringValue content = 4;
    google.protobuf.Struct metadata = 5;
    google.protobuf.BytesValue hash = 6;
  }
  repeated Document documents = 1;
}
message Embedding {
  message Embedding {
    google.protobuf.Struct attributes = 1;
    repeated float vector = 2;
    google.protobuf.StringValue text = 3;
  }
  repeated Embedding embeddings = 1;
  google.protobuf.StringValue model_name = 2;
}
message LLM {
  google.protobuf.Struct attributes = 1;
}
message Tool {
  google.protobuf.Struct attributes = 1;
}
